### PR TITLE
added polling for usage quota

### PR DESCRIPTION
**Pull request by NikitaPoly**

### CODE REVIEW

File: deployments/pulse3d/services/pulse3d/src/main.py [hunk 1]
```diff
@@ -769,3 +768,17 @@ async def get_versions(request: Request):
     except Exception as e:
         logger.error(f"Failed to retrieve info of pulse3d versions: {repr(e)}")
         raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+
+@app.get("/usage-quota", response_model=Union[UsageQuota, GenericErrorResponse])
```
Comment (luciipak) [inline deployments/pulse3d/services/pulse3d/src/main.py:773]: I would just call this `/usage`

Comment (luciipak) [reply deployments/pulse3d/services/pulse3d/src/main.py:773]: the `GenericErrorResponse` never gets returned here so I don't think it's necessary as a `response_model`

Comment (luciipak) [reply deployments/pulse3d/services/pulse3d/src/main.py:773]: and if this exists now, i don't know if it's necessary to check in the jobs and uploads routes

Comment (NikitaPoly) [reply deployments/pulse3d/services/pulse3d/src/main.py:773]: I think this is still necessary because those are the checks that will prevent an analysis if a job gets through the FE for a user that has used all his quota

File: frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js [hunk 2]
```diff
@@ -20,52 +20,69 @@ const ExpiredP = styled.p`
...
-
+  const [isExpired, setIsExpired] = useState();
+  const pollUsageQuota = async () => {
+    try {
+      const response = await fetch(`${process.env.NEXT_PUBLIC_PULSE3D_URL}/usage-quota`);
+      if (response && response.status === 200) {
+        const newUsageQuota = await response.json();
+        const limit = parseInt(newUsageQuota.limits.jobs);
+        const actual = parseInt(newUsageQuota.current.jobs);
+        const usagePercentage = Math.round((actual / limit) * 100);
```
Comment (luciipak) [inline frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js:35]: since usagePercentage is already being used as a variable name in state, I would name this something else so there isn't any unexpected behaviors

File: frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js [hunk 3]
```diff
@@ -20,52 +20,69 @@ const ExpiredP = styled.p`
...
+
+  return (
+    <>
+      {maxJobs === -1 && <ProgressDiv>Unlimited Access</ProgressDiv>}
+      {!isExpired && maxJobs !== -1 && (
         <ProgressDiv>
           <p>Usage</p>
           <CircularProgressWithLabel value={usagePercentage} colorOfTextLabel={colorOfTextLabel} />
-          <ProgressP>{`${actualUploads}/${maxUploads} Analysis used`}</ProgressP>
+          <ProgressP>{`${actualJobs}/${maxJobs} Analysis used`}</ProgressP>
```
Comment (luciipak) [inline frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js:82]: I'd make this `Analyses`

File: deployments/pulse3d/services/pulse3d/src/main.py (hunk 1 above)
Comment (luciipak) [reply deployments/pulse3d/services/pulse3d/src/main.py:773]: Eventually, we'll want to remove it from the jobs and uploads routes and just call this route before posting a new job etc and only call POST /jobs POST /uploads if this new route returns an unreached quota. But for now to get this out, we can just keep it.

File: deployments/pulse3d/services/pulse3d/src/main.py [hunk 4]
```diff
@@ -769,3 +768,17 @@ async def get_versions(request: Request):
...
+async def get_usage_quota(request: Request, token=Depends(ProtectedAny(scope=PULSE3D_USER_SCOPES))):
```
Comment (luciipak) [inline deployments/pulse3d/services/pulse3d/src/main.py:774]: I also think you'll want to add customer scopes here so that it is displayed for admin accounts

File: deployments/pulse3d/services/pulse3d/src/main.py [hunk 5]
```diff
@@ -769,3 +768,17 @@ async def get_versions(request: Request):
...
+    """Get the usage quota for the spesific user"""
```
Comment (luciipak) [inline deployments/pulse3d/services/pulse3d/src/main.py:775]: specific*

File: deployments/pulse3d/services/pulse3d/src/main.py [hunk 6]
```diff
@@ -769,3 +768,17 @@ async def get_versions(request: Request):
...
+    try:
+        customer_id = str(uuid.UUID(token["customer_id"]))
```
Comment (luciipak) [inline deployments/pulse3d/services/pulse3d/src/main.py:777]: and for customer accounts, this will be the `userid` instead of `customer_id`, so once these changes are made, doubel check this works for both account types

File: frontend/pulse3d/src/pages/upload-form.js [hunk 7]
```diff
@@ -176,9 +176,11 @@ export default function UploadForm() {
         tabSelection === "Re-analyze Existing Upload" &&
         usageQuota &&
         usageQuota.limits &&
-        parseInt(usageQuota.limits.jobs) !== -1
+        parseInt(usageQuota.limits.jobs) !== -1 &&
+        files.length > 0 &&
+        files[0].created_at !== files[0].updated_at
     );
-  }, [paramErrors, files, inProgress, wellGroupErr]);
+  }, [paramErrors, files, inProgress]);
```
Comment (luciipak) [inline frontend/pulse3d/src/pages/upload-form.js:183]: `wellGroupErr` needs to be here

File: deployments/pulse3d/services/pulse3d/src/main.py [hunk 8]
```diff
@@ -771,3 +770,23 @@ async def get_versions(request: Request):
     except Exception as e:
         logger.error(f"Failed to retrieve info of pulse3d versions: {repr(e)}")
         raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+
+@app.get("/usage", response_model=UsageQuota)
+async def get_usage_quota(
+    request: Request, token=Depends(ProtectedAny(scope=PULSE3D_SCOPES)), service: str = "pulse3d"
```
Comment (luciipak) [inline deployments/pulse3d/services/pulse3d/src/main.py:777]: this is an example how to use Query params 
https://github.com/CuriBio/k8s_curi/blob/4d110d238be88e7135cf96729d5adfd4d42cefca/deployments/pulse3d/services/pulse3d/src/main.py#L100

File: frontend/pulse3d/src/pages/upload-form.js [hunk 9]
```diff
@@ -176,7 +176,9 @@ export default function UploadForm() {
         tabSelection === "Re-analyze Existing Upload" &&
         usageQuota &&
         usageQuota.limits &&
```
Comment (luciipak) [inline frontend/pulse3d/src/pages/upload-form.js:178]: last nit picky comment, can we break up this big conditional into smaller ones with variable names that explain what it's checking? It's kind of hard to follow. Or at min, add some comments. How does this check if it's not the first reanalysis?

File: frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js [hunk 10]
```diff
@@ -20,52 +20,70 @@ const ExpiredP = styled.p`
...
-      usageState = (
-        <>
-          <ProgressDiv>Unlimited Access</ProgressDiv>
-        </>
-      );
-    } else if (!usageQuota.jobs_reached) {
-      // max usage has not been reached
-      usageState = (
+  useEffect(() => {
+    if (actualAnalyses) {
```
Comment (luciipak) [inline frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js:67]: is a new interval going to be started every time the actualAnalyses value changes?

Comment (NikitaPoly) [reply frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js:67]: I can confirm that the set interval is not being duplicated

Comment (NikitaPoly) [reply frontend/pulse3d/src/components/basicWidgets/UsageProgressWidget.js:67]: anytime the interval is started the previous one is cancelled

**Review by luciipak — APPROVED** [review]
